Watchman.Discord/Areas/Protection/Controllers/AntiSpamController.cs

```diff
@@ -1,44 +1,90 @@
         {
-            this._antiSpamService = antiSpamService;
-            _userMessagesCountService = userMessagesCountService;
-            this._strategy = new SpamDetectingStrategy();
+            _serverMessagesCacheService = serverMessagesCacheService;
+            _punishmentsCachingService = punishmentsCachingService;
+            _antiSpamService = antiSpamService;
+            _overallSpamDetector = OverallSpamDetectorStrategy.GetStrategyWithDefaultDetectors(serverMessagesCacheService, checkUserSafetyStrategyService);
+            _spamPunishmentStrategy = new SpamPunishmentStrategy(punishmentsCachingService);
         }
 
         [ReadAlways]
-        public Task Scan(DiscordRequest request, Contexts contexts)
+        public async Task Scan(DiscordRequest request, Contexts contexts)
         {
-            this._antiSpamService.AddUserMessage(contexts, request);
-            var messagesInShortTime = _antiSpamService.CountUserMessagesShorterTime(contexts.User.Id);
-            var messagesInLongTime = _antiSpamService.CountUserMessagesLongerTime(contexts.User.Id);
-            var userWarnsInLastFewMinutes = _antiSpamService.CountUserWarnsInShortTime(contexts.User.Id);
-            var userWarnsInLastFewHours = _antiSpamService.CountUserWarnsInLongTime(contexts.User.Id);
-            var userMutesInLastFewHours = _antiSpamService.CountUserMutesInLongTime(contexts.User.Id);
-            var userMessages = _userMessagesCountService.CountMessages(contexts.User.Id, contexts.Server.Id);
-
-            Log.Information($"Warns in few minutes: {userWarnsInLastFewMinutes}; in few hours: {userWarnsInLastFewHours}");
-            Log.Information($"Mutes in few hours: {userMutesInLastFewHours}");
-
-            var punishment = _strategy.SelectPunishment(userWarnsInLastFewMinutes, userWarnsInLastFewHours, userMutesInLastFewHours, messagesInShortTime, messagesInLongTime, userMessages);
-            _antiSpamService.SetPunishment(contexts, punishment);
+            var stopwatch = Stopwatch.StartNew();
+            Log.Information("Started scanning the message");
+
+            if (ShouldCheckThisMessage(contexts.User.Id, request.SentAt))
+            {
+                _isNowChecking = true;
+                var spamProbability = _overallSpamDetector.GetOverallSpamProbability(request, contexts);
+                if (spamProbability != SpamProbability.None)
+                {
+                    Log.Information("{SpamProbability} for {user}", spamProbability, contexts.User.Name);
+                    await HandlePossibleSpam(contexts, spamProbability, request.SentAt);
+                }
+                _isNowChecking = false;
+            }
+            _serverMessagesCacheService.AddMessage(request, contexts);
             Log.Information("Scanned");
-            return Task.CompletedTask;
+            Log.Information("antispam: {miliseconds}ms", stopwatch.ElapsedMilliseconds);
```

Comment:
myśle że lepsze będą ticks

Reply:
dobrze byłoby sie zastanowić nad ładowaniem statystyk do jakiejś bazy
żeby później łatwo analizować jak działa aplikacja, bez konieczności fizycznego zaglądania do niej
ale też żeby nie generowało to ogromnego obciążenia